pkg/controllers/raycluster_webhook.go

```diff
@@ -0,0 +1,134 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+
+	rayv1 "github.com/ray-project/kuberay/ray-operator/apis/ray/v1"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/utils/pointer"
+	ctrl "sigs.k8s.io/controller-runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+
+	"github.com/project-codeflare/codeflare-operator/pkg/config"
+)
+
+// log is for logging in this package.
+var rayclusterlog = logf.Log.WithName("raycluster-resource")
+
+func SetupRayClusterWebhookWithManager(mgr ctrl.Manager, cfg *config.KubeRayConfiguration) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(&rayv1.RayCluster{}).
+		WithDefaulter(&rayClusterDefaulter{
+			Config: cfg,
+		}).
+		Complete()
+}
+
+// +kubebuilder:webhook:path=/mutate-ray-io-v1-raycluster,mutating=true,failurePolicy=fail,sideEffects=None,groups=ray.io,resources=rayclusters,verbs=create;update,versions=v1,name=mraycluster.kb.io,admissionReviewVersions=v1
+
+type rayClusterDefaulter struct {
+	Config *config.KubeRayConfiguration
+}
+
+var _ webhook.CustomDefaulter = &rayClusterDefaulter{}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type
+func (r *rayClusterDefaulter) Default(ctx context.Context, obj runtime.Object) error {
+	raycluster := obj.(*rayv1.RayCluster)
+
+	if !pointer.BoolDeref(r.Config.RayDashboardOAuthEnabled, true) {
+		return nil
+	}
+
+	// Check and add OAuth proxy if it does not exist
+	for _, container := range raycluster.Spec.HeadGroupSpec.Template.Spec.Containers {
+		if container.Name == "oauth-proxy" {
+			rayclusterlog.V(2).Info("OAuth sidecar already exists, no patch needed")
+			return nil
+		}
+	}
+
+	rayclusterlog.V(2).Info("Adding OAuth sidecar container")
+	// definition of the new container
+	newOAuthSidecar := corev1.Container{
```

Comment:
Would it have sense to limit Resources for oauth container?

Reply:
Right, for the sake of keeping that PR from growing too big, I'd suggest to add that in a separate PR.